src/openstack_cloud/openstack_cloud.py

```diff
@@ -0,0 +1,608 @@
+            for address in network_addresses
+        ]
+
+        if not self.server_name.startswith(f"{prefix}-"):
+            # Should never happen.
+            raise ValueError(
+                f"Found openstack server {server.name} managed under prefix {prefix}, contact devs"
+            )
+        # Disable E203 (space before :) as it conflicts with the formatter (black).
+        self.instance_id = self.server_name[len(prefix) + 1 :]  # noqa: E203
+
+
+@contextmanager
+def _get_openstack_connection(
+    clouds_config: dict[str, dict], cloud: str
+) -> Iterator[OpenstackConnection]:
+    """Create a connection context managed object, to be used within with statements.
+
+    The file of _CLOUDS_YAML_PATH should only be modified by this function.
+
+    Args:
+        clouds_config: The configuration in clouds.yaml format to apply.
+        cloud: The name of cloud to use in the clouds.yaml.
+
+    Raises:
+        OpenStackError: if the credentials provided is not authorized.
+
+    Yields:
+        An openstack.connection.Connection object.
+    """
+    if not _CLOUDS_YAML_PATH.exists():
+        _CLOUDS_YAML_PATH.parent.mkdir(parents=True, exist_ok=True)
+        _CLOUDS_YAML_PATH.write_text(data=yaml.dump(clouds_config), encoding="utf-8")
+
+    # api documents that keystoneauth1.exceptions.MissingRequiredOptions can be raised but
+    # I could not reproduce it. Therefore, no catch here for such exception.
+    try:
+        with openstack.connect(cloud=cloud) as conn:
+            conn.authorize()
+            yield conn
+    # pylint thinks this isn't an exception, but does inherit from Exception class.
+    except openstack.exceptions.HttpException as exc:  # pylint: disable=bad-exception-cause
+        logger.exception("OpenStack API call failure")
+        raise OpenStackError("Failed OpenStack API call") from exc
+
+
+class OpenstackCloud:
+    """Client to interact with OpenStack cloud.
+
+    The OpenStack server name is managed by this cloud. Caller refers to the instances via
+    instance_id. If the caller needs the server name, e.g., for logging, it can be queried with
+    get_server_name.
+    """
+
+    def __init__(self, clouds_config: dict[str, dict], cloud: str, prefix: str):
+        """Create the object.
+
+        Args:
+            clouds_config: The openstack clouds.yaml in dict format.
+            cloud: The name of cloud to use in the clouds.yaml.
+            prefix: Prefix attached to names of resource managed by this instance. Used for
+                identifying which resource belongs to this instance.
+        """
+        self._clouds_config = clouds_config
+        self._cloud = cloud
+        self.prefix = prefix
+
+    # Ignore "Too many arguments" as 6 args should be fine. Move to a dataclass if new args are
+    # added.
+    def launch_instance(  # pylint: disable=R0913
+        self, instance_id: str, image: str, flavor: str, network: str, userdata: str
```

Comment:
Perhaps we rename `userdata` to something like cloud_init or whatever to make it clear it is the script that gets run to start the runner